File: ios-app/CalendarioCiclismo/ViewModels/ChampionshipsViewModel.swift
import Foundation

/// ViewModel del Modo Campeonatos — equivalente a `js/campeonatos.js`.
@MainActor
@Observable
final class ChampionshipsViewModel {
    var countries: [ChampionshipCountry] = []
    var isLoading = false
    var error: String?
    var hasLoaded = false
    /// Claves `raceId#stage|final` de pruebas con resultados in-house (keepForWeb).
    /// Si una celda está aquí, su trofeo lleva a la pantalla NATIVA de resultados.
    var inhouseKeys: Set<String> = []
    /// Filtro local (Hoy/Todas/Pro/Masc/Fem), sin persistencia. Empieza en `today`
    /// durante su rango (24–28 jun, predeterminado); fuera de él, en `all`.
    var activeFilter: ChampionshipsConfig.Filter = ChampionshipsConfig.defaultFilter

    /// Países con al menos una prueba visible bajo el filtro activo.
    var displayCountries: [ChampionshipCountry] {
        countries.filter { !$0.visibleSlots(filter: activeFilter).isEmpty }
    }

    func load() async {
        isLoading = true
        error = nil
        hasLoaded = true
        do {
            countries = try await SupabaseService.shared.loadChampionships()
        } catch {
            if Task.isCancelled { isLoading = false; return }
            self.error = error.localizedDescription
        }
        isLoading = false
        // Resultados in-house (no bloqueante): refresca los trofeos a la pantalla
        // nativa cuando aparezcan (los CN suelen no tener in-house, pero un volcado
        // PDF/UCI puede activarlo). El grid ya está pintado; esto solo reetiqueta.
        let raceIds = countries.flatMap { $0.slots.values.compactMap { $0.race?.id } }
        if !raceIds.isEmpty {
            inhouseKeys = await SupabaseService.shared.inhouseStageKeys(raceIds: raceIds)
        }
    }
}

File: android-app/app/src/main/java/app/calendariociclismo/android/ui/championships/ChampionshipsViewModel.kt
package app.calendariociclismo.android.ui.championships

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import app.calendariociclismo.android.data.model.ChampionshipCountry
import app.calendariociclismo.android.data.repository.CalendarRepository
import app.calendariociclismo.android.util.ChampionshipsConfig
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel del Modo Campeonatos — equivalente a `js/campeonatos.js`.
 * Carga la rejilla una vez; el filtro (Todas/Pro/Masc/Fem) es estado local
 * sin persistencia.
 */
class ChampionshipsViewModel(
    private val repo: CalendarRepository,
) : ViewModel() {

    data class State(
        val isLoading: Boolean = true,
        val error: String? = null,
        val countries: List<ChampionshipCountry> = emptyList(),
        val filter: ChampionshipsConfig.Filter = ChampionshipsConfig.defaultFilter(),
        /** Claves `raceId#stage|final` de pruebas con resultados in-house (keepForWeb):
         *  su trofeo lleva a la pantalla NATIVA de resultados. */
        val inhouseKeys: Set<String> = emptySet(),
    ) {
        /** Países con al menos una prueba visible bajo el filtro activo. */
        val displayCountries: List<ChampionshipCountry>
            get() = countries.filter { it.visibleSlots(filter).isNotEmpty() }
    }

    private val _state = MutableStateFlow(State())
    val state: StateFlow<State> = _state.asStateFlow()

    init { load() }

    fun setFilter(filter: ChampionshipsConfig.Filter) {
        _state.value = _state.value.copy(filter = filter)
    }

    fun load() {
        _state.value = _state.value.copy(isLoading = true, error = null)
        viewModelScope.launch {
            runCatching { repo.loadChampionships() }
                .onSuccess { countries ->
                    _state.value = _state.value.copy(isLoading = false, countries = countries, error = null)
                    // Resultados in-house (no bloqueante): reetiqueta los trofeos a la
                    // pantalla nativa cuando aparezcan (un volcado PDF/UCI puede activarlo).
                    val raceIds = countries.flatMap { c -> c.slots.values.mapNotNull { it.race?.id } }
                    if (raceIds.isNotEmpty()) {
                        val keys = repo.inhouseStageKeys(raceIds)
                        if (keys.isNotEmpty()) _state.value = _state.value.copy(inhouseKeys = keys)
                    }
                }
                .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message ?: "") }
        }
    }
}

class ChampionshipsViewModelFactory(
    private val repo: CalendarRepository,
) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return ChampionshipsViewModel(repo) as T
    }
}
